### DISCUSSION TITLE
Should use of Serial all be in one FreeRTOS task?

### DISCUSSION THREAD

**Discussion by fletchjp** [feilipu/Arduino_FreeRTOS_Library#133] [Q&A]
I am developing examples using Arduino FreeRTOS.

I make quite a lot of use of the Serial port for output and some input. Should I put all of this activity into a single Task?

John

**Comment by feilipu** [comment]
There is no obligation to use Serial in just one Task, as a). I believe the Serial Port instantiation happens only once, and b). it is useful to have output messages from multiple Tasks to illustrate the example code you're creating.

It is also a useful example to see how Serial output is handled. AFAIK, writing to the Serial transmitter buffer is done with the interrupt disabled, as I've never experienced interleaved outputs. But, I've also never aggressively tested this.

The [standard FreeRTOS test suite](https://github.com/FreeRTOS/FreeRTOS/tree/main/FreeRTOS/Demo/Common/Minimal) has a centralised error messaging system using stream buffers [IIRC](https://github.com/feilipu/z88dk-libraries/blob/master/freertos/demo/common/main.c#L30), but this method is just done as a further example.

**Reply by feilipu** [reply]
Just to add to that, I guess the application is for model railway control systems.

Some time ago I was writing [PID code to guide a small robot](https://feilipu.me/2011/05/02/dogbot-post-6-back-on-track/). I think the code could be helpful for speed control on trains, where the load (number of carriages, track voltage drop, engine temperature, etc) could be variable but train speed should be constant. The code is [in here](https://github.com/feilipu/avrfreertos/tree/master/DogBot).

**Reply by fletchjp** [reply]
Phillip

Thank you very much for this. I am just learning FreeRTOS.  I have been doing a lot of work with TaskManagerIO.

You are correct that I am involved with model railways. What you did not know is that I am a retired academic chemical engineer and for many years taught students about PID.  I have located the code and will look at it with interest.

I am attempting to develop codes which will run on a MEGA (a UNO is too small) and also in a DUE, where I have another version of FreeRTOS based on version 10. Your code is using version 11, so I am exploring the subtle differences.

I have now downloaded the code from avrfreertos and have a lot to look at.

Thank you again.

John

**Reply by fletchjp** [reply]
Hi.  I have had a look at your PID calculation. My guess is that you have played around with this to set the constants. It is done in integers which will change the answers. Also I cannot find any initial value for the integral term iError.

John

**Reply by feilipu** [reply]
The "[Dogbot](https://feilipu.me/tag/dogbot/)" was my first major foray into programming, and it was a far too complex project for a first step.  I managed to get pieces of it working, but I could never get it complete. It still sits in a bottom drawer as a reminder of my hubris.

Anyway, the `iError` initial value is set to 0 during initial instantiation of the structure. I learned PID theory at Uni, but never used it until the Dogbot project. So it was quite exhilarating to finally get it to run in a straight line, in the face of the adversity of real-life friction and drag.

Another thought...

If you're using a ATmega2560 (MEGA) then you might find additional RAM useful, particularly if you're using a DUE as an alternate as it has 96kB RAM available. I've had good experience with the MegaRAM and [QuadRAM](https://www.rugged-circuits.com/new-products/quadram) products. I preferred the MegaRAM as it used to be cheaper and sufficient for my needs. Code for ATmega2560 [EXT RAM support is here](https://github.com/feilipu/avrfreertos/blob/master/freeRTOS11xx/lib_ext_ram/ext_ram.c).
